File: backend/src/inf_canvas/canvas/reducer.py
"""Pure reducer over the Canvas Command protocol (mirror of reducer.ts).

`apply_command` returns a new CanvasState and never mutates its input.
"""

from ..schema.canvas import CanvasEdge, CanvasGroup, CanvasNode, CanvasState
from ..schema.commands import (
    AddGroupCommand,
    AddNodeCommand,
    BatchCommand,
    CanvasCommand,
    ClearCommand,
    ConnectCommand,
    DisconnectCommand,
    MoveNodeCommand,
    RemoveGroupCommand,
    RemoveNodeCommand,
    SelectCommand,
    UpdateEdgeCommand,
    UpdateGroupCommand,
    UpdateNodeCommand,
)


def apply_command(state: CanvasState, command: CanvasCommand) -> CanvasState:
    match command:
        case AddNodeCommand():
            if any(n.id == command.id for n in state.nodes):
                return state
            node = CanvasNode(
                id=command.id,
                type=command.equipment,
                position=command.position,
                label=command.label,
                rotation=command.rotation,
                data=command.data,
            )
            return state.model_copy(update={"nodes": [*state.nodes, node]})

        case UpdateNodeCommand():
            nodes = []
            for n in state.nodes:
                if n.id != command.id:
                    nodes.append(n)
                    continue
                patch = command.patch
                merged_data = n.data
                if patch.data is not None:
                    merged_data = {**(n.data or {}), **patch.data}
                nodes.append(
                    n.model_copy(
                        update={
                            "label": patch.label if patch.label is not None else n.label,
                            "rotation": patch.rotation
                            if patch.rotation is not None
                            else n.rotation,
                            "data": merged_data,
                        }
                    )
                )
            return state.model_copy(update={"nodes": nodes})

        case MoveNodeCommand():
            nodes = [
                n.model_copy(update={"position": command.position}) if n.id == command.id else n
                for n in state.nodes
            ]
            return state.model_copy(update={"nodes": nodes})

        case RemoveNodeCommand():
            return state.model_copy(
                update={
                    "nodes": [n for n in state.nodes if n.id != command.id],
                    "edges": [
                        e for e in state.edges if e.source != command.id and e.target != command.id
                    ],
                    "groups": [
                        g.model_copy(
                            update={"memberIds": [m for m in g.memberIds if m != command.id]}
                        )
                        for g in state.groups
                    ],
                }
            )

        case ConnectCommand():
            if any(e.id == command.id for e in state.edges):
                return state
            edge = CanvasEdge(
                id=command.id,
                source=command.source,
                sourcePort=command.sourcePort,
                target=command.target,
                targetPort=command.targetPort,
                data=command.data,
            )
            return state.model_copy(update={"edges": [*state.edges, edge]})

        case UpdateEdgeCommand():
            edges = [
                e.model_copy(update={"data": command.patch}) if e.id == command.id else e
                for e in state.edges
            ]
            return state.model_copy(update={"edges": edges})

        case DisconnectCommand():
            return state.model_copy(
                update={"edges": [e for e in state.edges if e.id != command.id]}
            )

        case ClearCommand():
            return state.model_copy(update={"nodes": [], "edges": [], "groups": []})

        case SelectCommand():
            # Selection is a client-side concern; no persisted change.
            return state

        case AddGroupCommand():
            if any(g.id == command.id for g in state.groups):
                return state
            group = CanvasGroup(
                id=command.id,
                label=command.label,
                position=command.position,
                width=command.width,
                height=command.height,
                memberIds=command.memberIds,
                color=command.color,
            )
            return state.model_copy(update={"groups": [*state.groups, group]})

        case UpdateGroupCommand():
            group_patch = {k: v for k, v in command.patch if v is not None}
            groups = [
                g.model_copy(update=group_patch) if g.id == command.id else g
                for g in state.groups
            ]
            return state.model_copy(update={"groups": groups})

        case RemoveGroupCommand():
            return state.model_copy(
                update={"groups": [g for g in state.groups if g.id != command.id]}
            )

        case BatchCommand():
            new_state = state
            for sub in command.commands:
                new_state = apply_command(new_state, sub)
            return new_state

    return state

File: frontend/src/schema/reducer.ts
/**
 * Pure reducer over the Canvas Command protocol.
 *
 * `applyCommand` returns a NEW CanvasState (never mutates its input), which
 * keeps it trivially testable and CRDT-friendly later. The backend mirrors this
 * logic in Python; keep the two in sync.
 */

import type { CanvasEdge, CanvasGroup, CanvasNode, CanvasState } from './canvas';
import type { CanvasCommand } from './commands';

function replaceNode(
  nodes: CanvasNode[],
  id: string,
  fn: (n: CanvasNode) => CanvasNode,
): CanvasNode[] {
  return nodes.map((n) => (n.id === id ? fn(n) : n));
}

export function applyCommand(state: CanvasState, command: CanvasCommand): CanvasState {
  switch (command.op) {
    case 'add_node': {
      if (state.nodes.some((n) => n.id === command.id)) return state;
      const node: CanvasNode = {
        id: command.id,
        type: command.equipment,
        position: command.position,
        ...(command.label !== undefined ? { label: command.label } : {}),
        ...(command.rotation !== undefined ? { rotation: command.rotation } : {}),
        ...(command.data !== undefined ? { data: command.data } : {}),
      };
      return { ...state, nodes: [...state.nodes, node] };
    }

    case 'update_node': {
      return {
        ...state,
        nodes: replaceNode(state.nodes, command.id, (n) => ({
          ...n,
          ...(command.patch.label !== undefined ? { label: command.patch.label } : {}),
          ...(command.patch.rotation !== undefined ? { rotation: command.patch.rotation } : {}),
          ...(command.patch.data !== undefined
            ? { data: { ...n.data, ...command.patch.data } }
            : {}),
        })),
      };
    }

    case 'move_node': {
      return {
        ...state,
        nodes: replaceNode(state.nodes, command.id, (n) => ({ ...n, position: command.position })),
      };
    }

    case 'remove_node': {
      return {
        ...state,
        nodes: state.nodes.filter((n) => n.id !== command.id),
        edges: state.edges.filter((e) => e.source !== command.id && e.target !== command.id),
        groups: state.groups.map((g) => ({
          ...g,
          memberIds: g.memberIds.filter((mid) => mid !== command.id),
        })),
      };
    }

    case 'connect': {
      if (state.edges.some((e) => e.id === command.id)) return state;
      const edge: CanvasEdge = {
        id: command.id,
        source: command.source,
        sourcePort: command.sourcePort,
        target: command.target,
        targetPort: command.targetPort,
        ...(command.data !== undefined ? { data: command.data } : {}),
      };
      return { ...state, edges: [...state.edges, edge] };
    }

    case 'update_edge': {
      return {
        ...state,
        edges: state.edges.map((e) =>
          e.id === command.id ? { ...e, data: { ...e.data, ...command.patch } } : e,
        ),
      };
    }

    case 'disconnect': {
      return { ...state, edges: state.edges.filter((e) => e.id !== command.id) };
    }

    case 'select':
    case 'clear': {
      // `select` is a transient UI concern handled by the store; `clear` is
      // handled where a fresh canvas is needed. Neither mutates persisted graph
      // data here.
      if (command.op === 'clear') {
        return { ...state, nodes: [], edges: [], groups: [] };
      }
      return state;
    }

    case 'add_group': {
      if (state.groups.some((g) => g.id === command.id)) return state;
      const group: CanvasGroup = {
        id: command.id,
        label: command.label,
        position: command.position,
        width: command.width,
        height: command.height,
        memberIds: command.memberIds,
        ...(command.color !== undefined ? { color: command.color } : {}),
      };
      return { ...state, groups: [...state.groups, group] };
    }

    case 'update_group': {
      return {
        ...state,
        groups: state.groups.map((g) =>
          g.id === command.id ? { ...g, ...command.patch } : g,
        ),
      };
    }

    case 'remove_group': {
      return { ...state, groups: state.groups.filter((g) => g.id !== command.id) };
    }

    case 'batch': {
      return command.commands.reduce(applyCommand, state);
    }

    default: {
      // Exhaustiveness guard — a new command op without a case fails the build.
      const _exhaustive: never = command;
      return _exhaustive;
    }
  }
}
